File: src/main/scala/com/dyonovan/neotech/managers/RecipeManager.scala
package com.dyonovan.neotech.managers

import java.util

import com.dyonovan.neotech.registries._
import net.minecraft.command.ServerCommandManager

/**
  * This file was created for NeoTech
  *
  * NeoTech is licensed under the
  * Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License:
  * http://creativecommons.org/licenses/by-nc-sa/4.0/
  *
  * @author Paul Davis <pauljoda>
  * @since 2/20/2016
  */
object RecipeManager {
    sealed trait RecipeType { def name : String }
    case object FluidFuels extends RecipeType { val name = "fluidfuels" }
    case object Crucible extends RecipeType { val name = "crucible" }
    case object Solidifier extends RecipeType { val name = "solidifier" }
    case object Alloyer extends RecipeType { val name = "alloyer" }

    lazy val recipeHandlers = new util.HashMap[RecipeType, AbstractRecipeHandler[_, _, _]]()

    def preInit() : Unit = {
        CraftingRecipeManager.preInit()
    }

    def init() : Unit = {
        recipeHandlers.put(Crucible,   new CrucibleRecipeManager().loadHandler())
        recipeHandlers.put(Solidifier, new SolidifierRecipeManager().loadHandler())
        recipeHandlers.put(FluidFuels, new FluidFuelRecipeHandler().loadHandler())
        recipeHandlers.put(Alloyer,    new AlloyerRecipeHandler().loadHandler())
        CrusherRecipeRegistry.init()
    }

    def initCommands(manager : ServerCommandManager) : Unit = {
        for(handlerType <- recipeHandlers.keySet().toArray) {
            val command = recipeHandlers.get(handlerType.asInstanceOf[RecipeType]).getCommand
            if(command != null)
                manager.registerCommand(command)
        }
    }

    /**
      * Used to get a handler Type
      *
      * @param handlerType The handler Type
      * @tparam H The Handler Object
      * @return The Handler
      */
    def getHandler[H <: AbstractRecipeHandler[_, _, _]](handlerType : RecipeType) : H =
        recipeHandlers.get(handlerType).asInstanceOf[H]

    /**
      * Used to get Handles by name
      *
      * @param name The name of the handler, eg "Crusher"
      * @tparam H The recipe handler type
      * @return The recipe Handler
      */
    def getHandler[H <: AbstractRecipeHandler[_, _, _]](name : String)  : Option[H] = {
        for(x <- recipeHandlers.keySet().toArray) {
            val handler = x.asInstanceOf[RecipeType]
            if(handler.name.equalsIgnoreCase(name))
                return Option(recipeHandlers.get(handler).asInstanceOf[H])
        }
        None
    }
}

File: src/main/java/com/teambrmodding/neotech/managers/RecipeManager.java
package com.teambrmodding.neotech.managers;

import com.teambrmodding.neotech.registries.*;
import gnu.trove.map.hash.THashMap;
import net.minecraft.command.CommandBase;
import net.minecraft.command.ServerCommandManager;

/**
 * This file was created for NeoTech
 *
 * NeoTech is licensed under the
 * Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License:
 * http://creativecommons.org/licenses/by-nc-sa/4.0/
 *
 * @author Paul Davis - pauljoda
 * @since 2/15/2017
 */
public class RecipeManager {

    // The list of all recipe handlers
    public static THashMap<RecipeType, AbstractRecipeHandler> recipeHandlers = new THashMap<>();

    /**
     * Called to register things
     */
    public static void preInit() {
        CraftingRecipeManager.preInit();
    }

    /**
     * Builds the recipe registry
     */
    public static void init() {
        recipeHandlers.put(RecipeType.FLUID_FUELS, new FluidFuelRecipeHandler().loadHandler());
        recipeHandlers.put(RecipeType.CRUSHER,     new CrusherRecipeHandler().loadHandler());
        recipeHandlers.put(RecipeType.CRUCIBLE,    new CrucibleRecipeHandler().loadHandler());
        recipeHandlers.put(RecipeType.SOLIDIFIER,  new SolidifierRecipeHandler().loadHandler());
        recipeHandlers.put(RecipeType.ALLOYER,     new AlloyerRecipeHandler().loadHandler());
        recipeHandlers.put(RecipeType.CENTRIFUGE,  new CentrifugeRecipeHandler().loadHandler());
    }

    /**
     * Loads all the commands from the handlers
     * @param manager The manager to register to
     */
    public static void initCommands(ServerCommandManager manager) {
        for(RecipeType recipeType : recipeHandlers.keySet()) {
            CommandBase command = recipeHandlers.get(recipeType).getCommand();
            if(command != null)
                manager.registerCommand(command);
        }
    }

    /*******************************************************************************************************************
     * Helper Methods                                                                                                  *
     *******************************************************************************************************************/

    /**
     * Used to get a recipe handler
     * @param recipeType The recipe type
     * @return The recipe handler
     */
    public static  <H extends AbstractRecipeHandler> H getHandler(RecipeType recipeType) {
        return (H) recipeHandlers.get(recipeType);
    }

    public enum RecipeType {
        FLUID_FUELS("fluid_fuels"),
        CRUSHER("crusher"),
        CRUCIBLE("crucible"),
        SOLIDIFIER("solidifier"),
        ALLOYER("alloyer"),
        CENTRIFUGE("centrifuge");

        private String name;

        RecipeType(String name) {
            this.name = name;
        }

        /**
         * Get the name of this recipe type
         * @return The name
         */
        public String getName() {
            return name;
        }
    }
}
